features.ts

```diff
@@ -0,0 +1,390 @@
+      const sceneIDs = queueDoc?.scene_ids ?? [];
+      if (sceneIDs.length === 0) {
+        res.status(500).json({
+          error: true,
+          message: "No scene present in queue",
+        });
+        return;
+      }
+
+      const scenes: Record<string, any>[] = [];
+      for (const id of sceneIDs) {
+        const scene = await state.scenes.findOne({ "_id": new ObjectId(id) });
+        if (scene === null) {
+          throw new Error(`Database consistency failure, feature queue missing scene ${id}`);
+        }
+        const sceneJson = await sceneToJson(scene, state, req.session);
+        scenes.push(sceneJson);
+      }
+
+      res.json({
+        error: false,
+        scenes
+      });
+
+    });
+
+  state.app.post(
+    "/features/queue",
+    async (req: JwtRequest, res: Response) => {
+      const ids = req.body.scene_ids;
+      const valid = Array.isArray(ids) && ids.every(id => typeof id === "string");
+      if (!valid) {
+        res.status(400).json({
+          error: true,
+          message: "Scene IDs should be an array of strings",
+        });
+        return;
+      }
+
+      const objectIDs = ids.map((id: string) => new ObjectId(id));
+      const scenes = await state.scenes.find({ "_id": { "$in" : objectIDs } }).toArray();
+      if (scenes.length !== objectIDs.length) {
+        res.status(400).json({
+          error: true,
+          message: "At least one of the scene IDs does not correspond to a valid scene",
+        });
+        return;
+      }
+
+      const result = await state.featureQueue.updateOne(
+        { queue: true },
+        { "$set": { scene_ids: objectIDs } },
+      );
+
+      if (result.modifiedCount === 1) {
+        res.json({
+          error: false,
+          message: "Queue updated successfully"
+        });
+      } else {
+        res.status(500).json({
+          error: true,
+          message: "Error updating queue in database"
+        });
+      }
+    });
+
+  state.app.get(
+    "/features/queue/next",
+    async (req: JwtRequest, res: Response) => {
+      const queueDoc = await state.featureQueue.findOne({ queue: true });
+      const sceneIDs = queueDoc?.scene_ids ?? [];
+      if (sceneIDs.length === 0) {
+        res.status(500).json({
+          error: true,
+          message: "No scenes present in queue",
+        });
+        return;
+      }
+      const id = sceneIDs[0];
+      
+      const scene = await state.scenes.findOne({ "_id": new ObjectId(id) });
+      if (scene === null) {
+        throw new Error(`Database consistency failure, feature queue missing scene ${id}`);
+      }
+      const sceneJson = await sceneToJson(scene, state, req.session);
+
+      res.json({
+        error: false,
+        scene: sceneJson,
+      });
+
+    });
+
+  state.app.post(
+    "/features/queue/pop",
+    async (req: JwtRequest, res: Response) => {
+      const result = await tryPopFromFeatureQueue(state);
```

Comment:
As per my comment on the `tryPopFromFeatureQueue` function, I am not sure if this endpoint will function as intended. It seems like we need to fetch the queue doc, save the 0'th ID, and then update the doc to remove that ID from the list.

(This would have busted semantics if multiple people are trying to update the doc at once. I'm not worried about that corner case right now, but if Mongo/Cosmos offers a true "pop" that can remove an item from the list *and return it to us all in one atomic operation*, that would be better.)

Reply:
It's worth noting that the operation actually is atomic (per the docs [here](https://mongodb.github.io/node-mongodb-native/6.3/classes/Collection.html#findOneAndUpdate)). Though we are missing any kind of write-lock handling, so there is still some future work to be done.